src/transpilation/transformers.ts

```diff
@@ -86,40 +86,51 @@ function loadTransformersFromOptions(program: ts.Program, diagnostics: ts.Diagno
     };
 
     const options = program.getCompilerOptions() as CompilerOptions;
-    if (!options.plugins) return customTransformers;
-
-    for (const [index, transformerImport] of options.plugins.entries()) {
-        if (!("transform" in transformerImport)) continue;
-        const optionName = `compilerOptions.plugins[${index}]`;
+    if (options.plugins) {
+        for (const [index, transformerImport] of options.plugins.entries()) {
+            if (!("transform" in transformerImport)) continue;
+            const optionName = `compilerOptions.plugins[${index}]`;
+
+            const { error: resolveError, result: factory } = resolvePlugin(
+                "transformer",
+                `${optionName}.transform`,
+                getConfigDirectory(options),
+                transformerImport.transform,
+                transformerImport.import
+            );
 
-        const { error: resolveError, result: factory } = resolvePlugin(
-            "transformer",
-            `${optionName}.transform`,
-            getConfigDirectory(options),
-            transformerImport.transform,
-            transformerImport.import
-        );
+            if (resolveError) diagnostics.push(resolveError);
+            if (factory === undefined) continue;
 
-        if (resolveError) diagnostics.push(resolveError);
-        if (factory === undefined) continue;
+            const { error: loadError, transformer } = loadTransformer(optionName, program, factory, transformerImport);
+            if (loadError) diagnostics.push(loadError);
+            if (transformer === undefined) continue;
 
-        const { error: loadError, transformer } = loadTransformer(optionName, program, factory, transformerImport);
-        if (loadError) diagnostics.push(loadError);
-        if (transformer === undefined) continue;
+            if (transformer.before) {
+                customTransformers.before.push(transformer.before);
+            }
 
-        if (transformer.before) {
-            customTransformers.before.push(transformer.before);
-        }
+            if (transformer.after) {
+                customTransformers.after.push(transformer.after);
+            }
 
-        if (transformer.after) {
-            customTransformers.after.push(transformer.after);
-        }
-
-        if (transformer.afterDeclarations) {
-            customTransformers.afterDeclarations.push(transformer.afterDeclarations);
+            if (transformer.afterDeclarations) {
+                customTransformers.afterDeclarations.push(transformer.afterDeclarations);
+            }
         }
     }
-
+    if (options.jsx === ts.JsxEmit.React) {
+        customTransformers.before.push(context => {
+            const patchedContext: ts.TransformationContext = {
+                ...context,
+                getCompilerOptions: () => ({
+                    ...context.getCompilerOptions(),
+                    target: ts.ScriptTarget.ESNext,
```

Comment:
why is it required to overwrite compiler options here? could you add a comment explaining the reasoning?

Reply:
If the script target is < ES.2017, typescript does some additional unnecessary transformations relating to object spreading. I override it here so we always get the latest ES version features